### PR TITLE
4.0.0 major release

### PR DESCRIPTION
**Pull request by mikechoch**
* Added a `setAppId` method, which will now be 1 of 2 important methods for OneSignal setup
* The goal is to have 2 necessary methods, `setAppId` and `setAppContext` and once these are called, the SDK will be initialized correctly

**NOTES:**
All major release PRs should be off of `major_release_4.0.0` branch
May need to rebase as things are merged into `major_release_4.0.0`

## 
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/912)

### CODE REVIEW

**Comment by mikechoch** [comment]
Planning to merge this branch after the 4.0.0 release features are finished and all merged here.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java [hunk 1]
```diff
@@ -2454,95 +2427,64 @@ private static long getLastSessionTime() {
...
-    * {@link OneSignal.OSInFocusDisplayOption#Notification Notification} - native notification display while user has app in focus (can be distracting).
-    * <br/>
-    * {@link OneSignal.OSInFocusDisplayOption#InAppAlert In-App Alert (Default)} - native alert dialog display, which can be helpful during development.
-    * <br/>
-    * {@link OneSignal.OSInFocusDisplayOption#None None} - notification is silent.
-    *
-    * @param displayOption the {@link OneSignal.OSInFocusDisplayOption OSInFocusDisplayOption} to set
+    * You can call this method with {@code false} to opt users out of receiving all notifications through
+    * OneSignal. You can pass {@code true} later to opt users back into notifications.
+    * @param disable whether to subscribe the user to notifications or not
```
Comment (rgomezp) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:2432]: this comment is wrong

Comment (jkasten2) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:2432]: ah, logic was flipped with the `setSubscription` to `dsiablePush` rename but this comment was missed.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java [hunk 2]
```diff
@@ -1690,21 +1553,20 @@ else if (keyValues.isNull(key) || "".equals(value)) {
             }
 
             if (!toSend.toString().equals("{}")) {
+               logger.debug("Available tags to send: " + toSend.toString());
```
Comment (rgomezp) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:1556]: Can we add better logging here? The else statement has "Send tags ended successfully"

Thus, what is the opposite of that? Why did the send tags not send successfully?

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java [hunk 3]
```diff
@@ -1604,10 +1441,22 @@ public static void removeExternalUserId(final OSExternalUserIdUpdateCompletionHa
     * @see OneSignal#deleteTag
     * @see OneSignal#deleteTags
     */
-   public static void sendTag(String key, String value) {
+   public static void sendTag(final String key, final String value) {
```
Comment (rgomezp) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:1444]: Can we add logging to all methods?

e.g:
```java
public static void sendTag(final String key, final String value) {
   logMethodName("sendTag");
   // ...
}
```

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java (hunk 2 above)
Comment (rgomezp) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:1556]: For context, I am falling into this codepath while testing RN and don't know why. Better logging here (and across all methods) would be helpful.

Logs:
```
2020-09-24 20:13:25.651 11163-11247/com.rnonesignal I/ReactNativeJS: Attempting to send tags
2020-09-24 20:13:25.672 11163-11248/com.rnonesignal D/OneSignal: Available tags to send: {"a":"1","b":"2"}
```

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java [hunk 4]
```diff
@@ -1915,95 +1777,35 @@ public static void deleteTags(Collection<String> keys, ChangeTagsUpdateHandler h
...
+      try {
+         JSONObject jsonTags = new JSONObject();
 
-      idsAvailableHandler.idsAvailable(userId, regId);
+         for (int i = 0; i < jsonArray.length(); i++)
+            jsonTags.put(jsonArray.getString(i), "");
 
-      if (regId != null)
-         idsAvailableHandler = null;
+         sendTags(jsonTags, handler);
```
Comment (rgomezp) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java:1805]: we should log something here to make clear why deleting tags results in a send tags call

**Pull request merged**